### PR TITLE
Support "special" search param Location-near #259

### PR DESCRIPTION
**Pull request by prb112**
Support "special" search param Location-near #259

This PR is broken into 3 core elements: 
1 - fhir-search location enablement 
2 - fhir-persistence-jdbc location enablement
3 - extraction of parameter values

    - Add support for 'special' in parseQueryString with Prefix
    - Add near.unitTest.properties to turn on fhir-search and
    fhir-persistence tracing
    - Update the AbstractQueryBuilder to wrap location parsing errors in a
    FHIRPersistenceException and re-throw
    - Add configuration for FHIRConfiguration for switching bounding type
    - Update the LocationParmBehaviorUtil to use BoundingType configuration
    - Add JDBCSearchNearTest
    - Update Units-of-Measure
    - Update FHIR Search NearLocationHandler
    - Update Bounding interface with Box, Radius types
    - Extend tests to cover the new types, and confirm validity of the types
    - Introduce enumeration with BoundingType
    - Update the QueryBuilder signature
    - Update NearLocationHandler
    - Refactor Units to use units-of-measure definitions
            - StatuteLengths
            - ImperialUnits
            - Metric Units
            - Remove OtherUnits (Angstrom)
            - configure Units in lookup map in UOMManager
    - Refactor FHIRPersistenceUtil to put NEAR processing into fhir-search
    - AbstractQueryBuilder
            - Change the boundingbox to a list of boundingradius
            - Change to checking near (near-distance is deprecated) and add to a
    check to NearLocationHandler to support a generic check
            - Add Bounding, BoundingBox, BoundingRadius
            - Add Tests for Bounding types
    - Update QuerySegmentAggregator to use the single check for near
    - Update JDBCQueryBuilder to match the changes to AbstractQueryBuilder
    - Introduce LocationBehaviorUtil to more consistently model the Location
    signature query
    - Remove from ParameterDAOImplementation near-distance
    - Add SearchConstants.Type 'special'
    - Add JDBCSearchNearTest with 'near'
    - Remove unused call to continue in SearchUtil
    - Update the validation for the BoundingBox
    - Remove the guard for the 'special' type which blocks near processing
    - Add support for the extraction routine in the JDBC visitor code
    (Location.position)
    - Add logicalId to the ParameterVisitorBatchDAO location
    - Add 'special' to the DTO Parameter object

### CODE REVIEW

File: docs/src/pages/guides/FHIRServerUsersGuide.md [hunk 1]
```diff
@@ -1212,6 +1212,7 @@ This section contains reference information about each of the configuration prop
 |`fhirServer/audit/serviceProperties/geoCity`|string|The Geo City configure for CADF audit logging service.|
 |`fhirServer/audit/serviceProperties/geoState`|string|The Geo State configure for CADF audit logging service.|
 |`fhirServer/audit/serviceProperties/geoCounty`|string|The Geo Country configure for CADF audit logging service.|
+|`fhirServer/search/boundingAreaType`|boolean|True, the bounding area is a Radius, else the bounding area is a box.|
```
Comment (lmsurpre) [inline docs/src/pages/guides/FHIRServerUsersGuide.md]: maybe rename it to `fhirServer/search/useBoundingRadius` to make it clear what true/false mean

or, alternatively, make it a string and validate that the value is either `box` or `radius`.

File: fhir-model/src/main/java/com/ibm/fhir/model/visitor/Visitor.java [hunk 2]
```diff
@@ -1347,4 +1348,14 @@
     void visit(java.lang.String elementName, Year value);
     void visit(java.lang.String elementName, YearMonth value);
     void visit(java.lang.String elementName, ZonedDateTime value);
+
+    /**
+     * @return the coordinates for the lat/long.
```
Comment (lmsurpre) [inline fhir-model/src/main/java/com/ibm/fhir/model/visitor/Visitor.java]: this isn't right.  it returns true or false which controls whether child elements should be visited.

File: fhir-model/src/main/java/com/ibm/fhir/model/visitor/Visitor.java [hunk 3]
```diff
@@ -1347,4 +1348,14 @@
     void visit(java.lang.String elementName, Year value);
     void visit(java.lang.String elementName, YearMonth value);
     void visit(java.lang.String elementName, ZonedDateTime value);
+
+    /**
+     * @return the coordinates for the lat/long. 
+     * 
+     * @param elementName
+     * @param elementIndex
+     * @param location
```
Comment (lmsurpre) [inline fhir-model/src/main/java/com/ibm/fhir/model/visitor/Visitor.java]: this should match the name of the parameter

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/dto/Parameter.java [hunk 4]
```diff
@@ -220,6 +220,9 @@ public void visit(IParameterVisitor visitor) throws FHIRPersistenceException {
         case URI:
             visitor.stringValue(name, this.valueString, baseLevel);
             break;
+        case SPECIAL:
```
Comment (lmsurpre) [inline fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/dto/Parameter.java]: i think there is one other parameter with type=special ...maybe we should verify that we have a 'near' search parameter here (just to be safe)

Comment (prb112) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/dto/Parameter.java]: I checked this out using the search-parameters.json and the website.

Comment (prb112) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/dto/Parameter.java]: only 1

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/type/LocationParmBehaviorUtil.java [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+import com.ibm.fhir.search.location.bounding.Bounding;
+import com.ibm.fhir.search.location.bounding.BoundingBox;
+import com.ibm.fhir.search.location.bounding.BoundingRadius;
+
+/**
+ * Location Behavior Util generates SQL and loads the variables into bind
+ * variables.
+ * <br>
+ * The bounding areas at this point are already valid as the fhir-search
+ * component has processed the received point into a set of bounding boxes.
```
Comment (lmsurpre) [inline fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/type/LocationParmBehaviorUtil.java]: is this right?  i see two buildLocationSearchQuery methods: one takes a List of Boundings (not boxes) and the other takes a `single` bounding box.

Comment (prb112) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/type/LocationParmBehaviorUtil.java]: There are actually three methods
1 - coordinates the type of the boundingarea 
2 - for the radius 
3 - for the box

File: fhir-persistence-jdbc/src/test/java/com/ibm/fhir/persistence/jdbc/search/test/JDBCSearchNearTest.java [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+        }
+
+        FHIRSearchContext ctx = SearchUtil.parseQueryParameters(Location.class, queryParms, true);
+        FHIRPersistenceContext persistenceContext = FHIRPersistenceContextFactory.createPersistenceContext(null, ctx);
+        MultiResourceResult<Resource> result = persistence.search(persistenceContext, Location.class);
+        return result;
+    }
+
+    @AfterClass
+    public void teardown() throws Exception {
```
Comment (lmsurpre) [inline fhir-persistence-jdbc/src/test/java/com/ibm/fhir/persistence/jdbc/search/test/JDBCSearchNearTest.java]: should we also delete the saved resource(s)?

Comment (lmsurpre) [inline fhir-persistence-jdbc/src/test/java/com/ibm/fhir/persistence/jdbc/search/test/JDBCSearchNearTest.java]: should we also delete the saved resource(s)?

Comment (prb112) [reply fhir-persistence-jdbc/src/test/java/com/ibm/fhir/persistence/jdbc/search/test/JDBCSearchNearTest.java]: sure ... that's no problem to add

File: fhir-persistence-jdbc/src/test/java/com/ibm/fhir/persistence/jdbc/search/test/JDBCSearchNearTest.java [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+public class JDBCSearchNearTest {
+    private Properties testProps;
+
+    protected Location savedResource;
+
+    protected static FHIRPersistence persistence = null;
+
+    @BeforeClass
+    public void startup() throws Exception {
+        InputStream inputStream = JDBCSearchNearTest.class.getResourceAsStream("/near.unitTest.properties");
```
Comment (lmsurpre) [inline fhir-persistence-jdbc/src/test/java/com/ibm/fhir/persistence/jdbc/search/test/JDBCSearchNearTest.java]: does it need its own unitTest properties file?  why not just re-use `logging.unitTest.properties`?

Comment (prb112) [reply fhir-persistence-jdbc/src/test/java/com/ibm/fhir/persistence/jdbc/search/test/JDBCSearchNearTest.java]: I turn on com.ibm.fhir.search trace with this one.

Comment (prb112) [reply fhir-persistence-jdbc/src/test/java/com/ibm/fhir/persistence/jdbc/search/test/JDBCSearchNearTest.java]: also exists in fhir-persistence-jdbc

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/dto/Parameter.java (hunk 4 above)
Comment (lmsurpre) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/dto/Parameter.java]: 2 listed at https://www.hl7.org/fhir/search.html#special
* near
* _filter

but yeah, if `_filter` doesn't appear in search-parameters.json then i guess we're good

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/type/LocationParmBehaviorUtil.java (hunk 5 above)
Comment (lmsurpre) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/type/LocationParmBehaviorUtil.java]: ok, i guess my point was that the "the fhir-search component has processed the received point into a set of bounding boxes" comment doesn't see to belong here (at the class level)...maybe move it to one of the methods or just delete it?

File: fhir-persistence-jdbc/src/test/java/com/ibm/fhir/persistence/jdbc/search/test/JDBCSearchNearTest.java (hunk 7 above)
Comment (lmsurpre) [reply fhir-persistence-jdbc/src/test/java/com/ibm/fhir/persistence/jdbc/search/test/JDBCSearchNearTest.java]: yeah, i guess i was just thinking you could add a line for `com.ibm.fhir.search` in the existing `logging.unitTest.properties` file.  both will be INFO by default and you can selectively change the log level.  i don't see why we need a separate file for it, but i guess its not a big deal.

Comment (prb112) [reply fhir-persistence-jdbc/src/test/java/com/ibm/fhir/persistence/jdbc/search/test/JDBCSearchNearTest.java]: right.  just explaining my choice.  I can reconcile with the existing one

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/dto/Parameter.java (hunk 4 above)
Comment (prb112) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/dto/Parameter.java]: it is not.

File: fhir-model/src/main/java/com/ibm/fhir/model/visitor/Visitor.java (hunk 2 above)
Comment (prb112) [reply fhir-model/src/main/java/com/ibm/fhir/model/visitor/Visitor.java]: fixed in code generator.  I modified this class by hand (as you can see above).  I have fixed this case.

File: fhir-model/src/main/java/com/ibm/fhir/model/visitor/Visitor.java (hunk 3 above)
Comment (prb112) [reply fhir-model/src/main/java/com/ibm/fhir/model/visitor/Visitor.java]: fixed in code generator.  I modified this class by hand (as you can see above).  I have fixed this case.

File: fhir-persistence-jdbc/src/test/java/com/ibm/fhir/persistence/jdbc/search/test/JDBCSearchNearTest.java (hunk 7 above)
Comment (prb112) [reply fhir-persistence-jdbc/src/test/java/com/ibm/fhir/persistence/jdbc/search/test/JDBCSearchNearTest.java]: Reconciled.

File: fhir-persistence-jdbc/src/test/java/com/ibm/fhir/persistence/jdbc/search/test/JDBCSearchNearTest.java (hunk 6 above)
Comment (prb112) [reply fhir-persistence-jdbc/src/test/java/com/ibm/fhir/persistence/jdbc/search/test/JDBCSearchNearTest.java]: done

File: docs/src/pages/guides/FHIRServerUsersGuide.md (hunk 1 above)
Comment (prb112) [reply docs/src/pages/guides/FHIRServerUsersGuide.md]: Fixed

Comment (prb112) [inline docs/src/pages/guides/FHIRServerUsersGuide.md]:
```suggestion
|`fhirServer/search/useBoundingRadius`|boolean|True, the bounding area is a Radius, else the bounding area is a box.|
```

File: fhir-model/src/main/java/com/ibm/fhir/model/visitor/DefaultVisitor.java [hunk 8]
```diff
@@ -29,8 +29,8 @@
  * <ul>
  * <li>preVisit methods to control whether a given Resource or Element gets visited
  * <li>visitStart methods to provide setup behavior prior to the visit
- * <li>supertype visit methods to perform some common action on all visited Resources and Elements
- * <li>subtype visit methods to perform unique behavior that varies by the type being visited
+ * <li>defaultAction methods to perform some common action on all visited Resources and Elements
+ * <li>specific visit methods to perform unique behavior that varies by the type being visited
```
Comment (lmsurpre) [inline fhir-model/src/main/java/com/ibm/fhir/model/visitor/DefaultVisitor.java]: I forgot this class was generated and I updated this javadoc by hand.  Regenerating the model has overwritten these edits.  Mind updating the codeGenerator to preserve the `supertype`/`subtype` version?

Comment (prb112) [reply fhir-model/src/main/java/com/ibm/fhir/model/visitor/DefaultVisitor.java]: got it.  updating the code generator

**Pull request merged**